### PR TITLE
Example for adaptive step size choice

### PR DESCRIPTION
**Pull request by ParticularlyPythonicBS**
PR to complete [https://github.com/patrick-kidger/diffrax/issues/371](https://github.com/patrick-kidger/diffrax/issues/371
)
Added example to PIDController with a simple pendulum example illustrating the effect of tolerance of adaptive step choice controller.

### CODE REVIEW

File: diffrax/_step_size_controller/adaptive.py [hunk 1]
```diff
@@ -179,6 +179,52 @@ class PIDController(
         common to refer to solving an equation to specific tolerances, without
         necessarily stating which solver was used.)
 
+        !!! Example
```
Comment (patrick-kidger) [inline diffrax/_step_size_controller/adaptive.py:182]: This is quite a long example, so let's maybe make it expandable with `???`

File: diffrax/_step_size_controller/adaptive.py [hunk 2]
```diff
@@ -179,6 +179,52 @@ class PIDController(
...
+                    domega = - jnp.sin(theta)
+                    return jnp.array([dtheta, domega])
+            dynamics = SimplePendulum()
+            init_angle = 0.1
+            init_omega = 0
+            state0 = jnp.array([init_angle, init_omega])
+            ```
+            We can integrate this using:
+            ```python
+            def integrator(dynamics, state0, stepsize_controller):
```
Comment (patrick-kidger) [inline diffrax/_step_size_controller/adaptive.py:201]: I think it'd be good to make this example a little more concise. Perhaps something like this:
```python
import diffrax as dfx

def dynamics(t, y, args):
    dtheta = y["omega"]
    domega = - jnp.sin(y["theta"])
    return dict(theta=dtheta, omega=domega)

y0 = dict(theta=0.1, omega=0)
term = dfx.ODETerm(dynamics)
sol = dfx.diffeqsolve(
    term, solver, t0=0, t1=1000, dt0=0.1, y0,
    saveat=dfx.SaveAts(ts=jnp.linspace(0, 1000, 10000),
    max_steps=2**20,
    stepsize_controller=...
)
```
WDYT?

(WIth a little `...`, as we discuss different choices of that below.)

File: diffrax/_step_size_controller/adaptive.py [hunk 3]
```diff
@@ -179,6 +179,52 @@ class PIDController(
...
+            to compare the effect of different tolerances:
+            ```python
+            PID_controller_incorrect = diffrax.PIDController(rtol=1e-3, atol=1e-6)
+            PID_controller_correct = diffrax.PIDController(rtol=1e-7, atol=1e-9)
+            Constant_controller = diffrax.ConstantStepSize()
+            ```
+            The phase portraits of the pendulum from the different tolerances clearly
+            illustrate the impact of the choice of `rtol` and `atol` on the accuracy of
+            the solution.
+            ![Phase portrait of pendulum](../imgs/pendulum_adaptive_steps.png)
```
Comment (patrick-kidger) [inline diffrax/_step_size_controller/adaptive.py:226]: I probably wouldn't use "incorrect" in the title here, as it's suggestive of a bug. This is the numerical method doing exactly what's expected, namely a cheap low-accuracy solution!
Maybe something like "less accurate" and "more accurate" instead?

**Review by patrick-kidger — COMMENTED** [review]
Thanks for the PR! I think this'd be a great addition to the docs.

File: diffrax/_step_size_controller/adaptive.py (hunk 1 above)
Comment (ParticularlyPythonicBS) [reply diffrax/_step_size_controller/adaptive.py:182]: That sounds great!

File: diffrax/_step_size_controller/adaptive.py (hunk 2 above)
Comment (ParticularlyPythonicBS) [reply diffrax/_step_size_controller/adaptive.py:201]: That works well, 
yeah the class was a vestige of the much larger system I simplified for this, but this is much better as an example

File: diffrax/_step_size_controller/adaptive.py (hunk 3 above)
Comment (ParticularlyPythonicBS) [reply diffrax/_step_size_controller/adaptive.py:226]: pendulum_adaptive_steps

Sounds good, here is the corrected plot.

**Comment by ParticularlyPythonicBS** [comment]
Just following up to see if anything else is needed

**Comment by patrick-kidger** [comment]
Can you push your changes from the last round of comments?
After that, let's merge this! :)

**Comment by patrick-kidger** [comment]
Alright, merged -- thank you for the contribution!
